go/felix/intdataplane/endpoint_mgr_test.go

```diff
@@ -0,0 +1,406 @@
+import (
+	log "github.com/Sirupsen/logrus"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/felix/go/felix/ipsets"
+	"github.com/projectcalico/felix/go/felix/iptables"
+	"github.com/projectcalico/felix/go/felix/proto"
+	"github.com/projectcalico/felix/go/felix/rules"
+	"github.com/projectcalico/felix/go/felix/set"
+	"reflect"
+)
+
+type mockTable struct {
+	expectedUpdate chan []*iptables.Chain
+	expectedRemove chan []*iptables.Chain
+
+	lastWorkDispatch []*iptables.Chain
+	lastHostDispatch []*iptables.Chain
+	lastHostEndpoint []*iptables.Chain
+}
+
+func newMockTable() *mockTable {
+	return &mockTable{
+		expectedUpdate: make(chan []*iptables.Chain, 2),
+		expectedRemove: make(chan []*iptables.Chain, 2),
+	}
+}
+
+func logChains(message string, chains []*iptables.Chain) {
+	if chains == nil {
+		log.Debug(message, " with nil chains")
+	} else {
+		log.WithField("chains", chains).Debug(message)
+		for _, chain := range chains {
+			log.WithField("chain", *chain).Debug("")
+		}
+	}
+}
+
+func (t *mockTable) UpdateChains(chains []*iptables.Chain) {
+	logChains("UpdateChains", chains)
+	expected := <-t.expectedUpdate
+	logChains("UpdateExpected", expected)
+	Expect(reflect.DeepEqual(chains, expected)).To(BeTrue())
+}
+
+func (t *mockTable) RemoveChains(chains []*iptables.Chain) {
+	logChains("RemoveChains", chains)
+	expected := <-t.expectedRemove
+	logChains("RemoveExpected", expected)
+	Expect(reflect.DeepEqual(chains, expected)).To(BeTrue())
+}
+
+func (t *mockTable) expectWorkDispatchChange(chains []*iptables.Chain) {
+	t.expectedRemove <- t.lastWorkDispatch
+	t.lastWorkDispatch = chains
+	t.expectedUpdate <- chains
+}
+
+func (t *mockTable) expectHostDispatchChange(chains []*iptables.Chain) {
+	t.expectedRemove <- t.lastHostDispatch
+	t.lastHostDispatch = chains
+	t.expectedUpdate <- chains
+}
+
+func (t *mockTable) expectHostEndpointChange(chains []*iptables.Chain) {
+	if t.lastHostEndpoint != nil {
+		t.expectedRemove <- t.lastHostEndpoint
+	}
+	t.lastHostEndpoint = chains
+	if t.lastHostEndpoint != nil {
+		t.expectedUpdate <- chains
+	}
+}
+
+var _ = Describe("EndpointManager test", func() {
+
+	var epMgr *endpointManager
+	var filterTable *mockTable
+
+	rrConfigNormal := rules.Config{
+		IPIPEnabled:          true,
+		IPIPTunnelAddress:    nil,
+		IPSetConfigV4:        ipsets.NewIPVersionConfig(ipsets.IPFamilyV4, "cali", nil, nil),
+		IPSetConfigV6:        ipsets.NewIPVersionConfig(ipsets.IPFamilyV6, "cali", nil, nil),
+		IptablesMarkAccept:   0x8,
+		IptablesMarkNextTier: 0x10,
+	}
+
+	for ip_version := range []uint8{4, 6} {
+		BeforeEach(func() {
+			renderer := rules.NewRenderer(rrConfigNormal)
+			filterTable = newMockTable()
+			epMgr = newEndpointManager(
+				filterTable,
+				renderer,
+				nil,
+				uint8(ip_version),
+				[]string{"cali"},
+				nil,
+			)
+		})
+
+		It("should be constructable", func() {
+			Expect(epMgr).ToNot(BeNil())
+		})
+
+		It("should process host endpoints", func() {
```

Comment:
I don't follow, the sequences of method calls to the object are the same.  The advantage is that it isolates each assertion into a separate test case.

For example, suppose you want to test two sequences of operations with assertions after operation:
```
A->B->C
```
and
```
A->B->D
```
you'd do
```
Describe("A", func() {
  BeforeEach(func() {
    A
  })
  
  It("should do the thing we expect in state A")

  Describe("then B", func() {
    BeforeEach(func() {
      B
    })

    It("should do the thing we expect in A->B", ...)
    It("should do the second thing we expect in A->B", ...)
    It("should do the third thing we expect in A->B", ...)

    Describe("then C", func() {
      BeforeEach(func() {
        C
      })
      It("should do the thing we expect in A->B->C", ...)
      ...
    })

    Describe("then D", func() {
      BeforeEach(func() {
        D
      })
      It("should do the thing we expect in A->B->D", ...)
      ...
    })
  })
})
```